backend/server.go

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package backend
+
+import "github.com/m3db/m3metrics/policy"
+
+// Server is a backend server that is responsible for encoding the
+// incoming metrics and the corresponding policies, and routing them to
+// the downstream service.
+type Server interface {
+	// Open opens the server.
+	Open() error
+
+	// WriteCounterWithPolicies writes a counter along with policies.
+	WriteCounterWithPolicies(id []byte, val int64, vp policy.VersionedPolicies) error
```

Comment:
`metric.ID` is an abstraction that represents a metric id that allows you to look up tag value based on a tag name, this is useful to handle say different stats protocols (e.g., m3 metric implements it differently from statsd, in the sense that m3 metric id is just gonna be a type alias for byte slice that implements the `TagValue` method on the `metric.ID` interface). Once we finish extracting the tag values and get to the server, all we care about are the underlying raw bytes, which is why the interface then takes raw byte slices.